dnf5daemon-server/callbacks.cpp

```diff
@@ -29,27 +29,44 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include <string>
 
 void * DownloadCallbacks::add_new_download(
-    void * user_data, [[maybe_unused]] const char * description, [[maybe_unused]] double total_to_download) {
+    void * user_data, const char * description, [[maybe_unused]] double total_to_download) {
+    if (auto * dbus_cb = static_cast<DbusCallback *>(user_data)) {
+        if (auto * dbus_repo_cb = dynamic_cast<DbusRepoCB *>(dbus_cb)) {
+            dbus_repo_cb->start(description);
+        }
+    }
     return user_data;
 }
 
 int DownloadCallbacks::progress(void * user_cb_data, double total_to_download, double downloaded) {
-    if (auto * dbus_package_ck = static_cast<DbusPackageCB *>(user_cb_data)) {
-        return dbus_package_ck->progress(total_to_download, downloaded, is_time_to_print());
+    if (auto * dbus_cb = static_cast<DbusCallback *>(user_cb_data)) {
+        if (auto * dbus_package_cb = dynamic_cast<DbusPackageCB *>(dbus_cb)) {
```

Comment:
Wouldn't it be better instead always doing those dynamic_cast just prepare both methods from libdnf::repo::DownloadCallbacks and from libdnf::repo::RepoCallbacks into the DbusCallback class and let the polymorphism do the talking instead of those ifs? For package callbacks the repo ones would be empty / not implemented in the related subclass and for repo callbacks the package ones. But I would say, it would be clearer and more readable...